src/storage/mod.rs

```diff
@@ -585,6 +587,69 @@ pub(crate) fn source_archive_path(name: &str, version: &str) -> String {
     format!("sources/{name}/{version}.zip")
 }
 
+#[instrument(skip(storage))]
+fn migrate_one(storage: &Storage, archive_path: &str) -> Result<()> {
+    // this will also download the index if it doesn't exist locally
+    let local_index_filename = storage.get_index_filename(archive_path)?;
+
+    if archive_index::is_sqlite_file(&local_index_filename)? {
+        info!("index already in SQLite format, skipping");
+        return Ok(());
+    }
+
+    info!("converting local index...");
+    let remote_index_path = format!("{}.index", &archive_path);
+    let new_index_temp_path = archive_index::convert_to_sqlite_index(&local_index_filename)?;
+
+    // first upload to S3, ongoing requests will still use the local CBOR index
+    info!("uplading to S3...");
+    // S3 put-object will overwrite the existing index
+    storage.store_one(remote_index_path, std::fs::read(&new_index_temp_path)?)?;
+
+    // move the temporary file into place
+    // This has a race condition when a request is trying to read the index between
+    // the `remove_file` and the `rename`. In this case the handler will then just
+    // unnecessarily download the index again.
+    fs::remove_file(&local_index_filename)?;
+    fs::rename(new_index_temp_path, local_index_filename)?;
+
+    Ok(())
+}
+
+/// migrate existing archive indexes from the old CBOR format to SQLite
+pub fn migrate_old_archive_indexes(
+    storage: &Storage,
+    conn: &mut impl postgres::GenericClient,
+) -> Result<()> {
+    for row in conn
+        .query_raw(
+            "
+            SELECT 
+                crates.name,
+                releases.version
+            FROM 
+                crates
+                INNER JOIN releases ON releases.crate_id = crates.id
+            WHERE 
+                releases.archive_storage = true
```

Comment:
This needs `build_status = true` as well, otherwise it attempts to migrate crates that failed to build.

(It may also pay to not fail if migrating one version fails, just log it and continue migrating more).

Reply:
> This needs `build_status = true` as well, otherwise it attempts to migrate crates that failed to build.

added 

> 
> (It may also pay to not fail if migrating one version fails, just log it and continue migrating more).

I added this too.